feat(baseapp): add option to discard events

aaronc opened the pull request:
# Description

This adds a `BaseApp` option to totally disable event emission. A follow-up to this PR would be adding an app.toml setting to configure this.

<!-- Add a description of the changes that this PR introduces and the files that
are the most critical to review. -->

---

## Author Checklist

*All items are required. Please add a note to the item if the item is not applicable and
please add links to any relevant follow up issues.*

I have...

* [ ] included the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title, you can find examples of the prefixes below:
    <!-- * `feat`: A new feature
    * `fix`: A bug fix
    * `docs`: Documentation only changes
    * `style`: Changes that do not affect the meaning of the code (white-space, formatting, missing semi-colons, etc)
    * `refactor`: A code change that neither fixes a bug nor adds a feature
    * `perf`: A code change that improves performance
    * `test`: Adding missing tests or correcting existing tests
    * `build`: Changes that affect the build system or external dependencies (example scopes: gulp, broccoli, npm)
    * `ci`: Changes to our CI configuration files and scripts (example scopes: Travis, Circle, BrowserStack, SauceLabs)
    * `chore`: Other changes that don't modify src or test files
    * `revert`: Reverts a previous commit -->
* [ ] confirmed `!` in the type prefix if API or client breaking change
* [ ] targeted the correct branch (see [PR Targeting](https://github.com/cosmos/cosmos-sdk/blob/main/CONTRIBUTING.md#pr-targeting))
* [ ] provided a link to the relevant issue or specification
* [ ] reviewed "Files changed" and left comments if necessary
* [ ] included the necessary unit and integration [tests](https://github.com/cosmos/cosmos-sdk/blob/main/CONTRIBUTING.md#testing)
* [ ] added a changelog entry to `CHANGELOG.md`
* [ ] updated the relevant documentation or specification, including comments for [documenting Go code](https://blog.golang.org/godoc)
* [ ] confirmed all CI checks have passed

## Reviewers Checklist

*All items are required. Please add a note if the item is not applicable and please add
your handle next to the items reviewed if you only reviewed selected items.*

Please see [Pull Request Reviewer section in the contributing guide](../CONTRIBUTING.md#reviewer) for more information on how to review a pull request.

I have...

* [ ] confirmed the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
* [ ] confirmed all author checklist items have been addressed
* [ ] reviewed state machine logic, API design and naming, documentation is accurate, tests and test coverage


<!-- This is an auto-generated comment: release notes by coderabbit.ai -->

## Summary by CodeRabbit

- **New Features**
  - Added a configurable option to disable event emission during block finalization and transaction processing, which can improve performance in high-throughput scenarios.

- **Refactor**
  - Streamlined event management by updating the routing mechanism for handling events, ensuring consistent and configurable behavior across processing contexts.

<!-- end of auto-generated comment: release notes by coderabbit.ai -->

technicallyty commented:
do we have an issue for the app.toml follow up?

aljo242 commented on baseapp/msg_service_router.go:
```diff
@@ -27,10 +27,11 @@ type MessageRouter interface {
+	routes               map[string]MsgServiceHandler
+	hybridHandlers       map[string]func(ctx context.Context, req, resp protoiface.MessageV1) error
+	circuitBreaker       CircuitBreaker
+	disableEventEmission bool
```
```suggestion
	discardEvents bool
```

for consistency

aljo242 commented on baseapp/msg_service_router.go:
```diff
@@ -215,6 +216,13 @@ func (msr *MsgServiceRouter) SetInterfaceRegistry(interfaceRegistry codectypes.I
 	msr.interfaceRegistry = interfaceRegistry
 }
 
+func (msr *MsgServiceRouter) newEventManager() sdk.EventManagerI {
```
we should godoc this

aljo242 commented on baseapp/discard_event_mgr.go:
```diff
@@ -0,0 +1,33 @@
+package baseapp
```
let's name this file `discard_event_manager.go`

aljo242 commented on baseapp/msg_service_router.go:
```diff
@@ -27,10 +27,11 @@ type MessageRouter interface {
-	routes            map[string]MsgServiceHandler
-	hybridHandlers    map[string]func(ctx context.Context, req, resp protoiface.MessageV1) error
-	circuitBreaker    CircuitBreaker
+	interfaceRegistry    codectypes.InterfaceRegistry
```
We write a test that demonstrates how this new option works:

- events are emitted when the field is false
- events are not emitted when true

by running some SDK message on the router

aljo242 commented on baseapp/options.go:
```diff
@@ -403,3 +403,9 @@ func (app *BaseApp) SetMsgServiceRouter(msgServiceRouter *MsgServiceRouter) {
+
+// SetDisableEventEmission configures whether or not events should be discarded whenever they are collected.
+// Disabling events on validator nodes will speed up block processing.
+func (app *BaseApp) SetDisableEventEmission(disableEventEmission bool) {
```
if we change everything to "discard" we should also change this

aaronc commented:
@aljo242 I've addressed your comments and also handled a few cases for emit emission that I originally missed. Let me know if you see anything else that needs to be addressed!

aaronc commented:
> do we have an issue for the app.toml follow up?

Not yet. We should add one. I looked quickly and maybe I'm missing something, but I'm not seeing a straightforward way to do this without adding it specifically to each app.go.

aljo242 commented:
Blocking until we can perform ironbird tests against this code
